crates/re_ws_comms/src/server.rs

```diff
@@ -121,201 +157,185 @@ impl RerunServer {
-            tokio::spawn(accept_connection(
-                log_stream.clone(),
-                peer,
-                tcp_stream,
-                history.clone(),
-            ));
+            }
         }
     }
 
-    /// Contains the `ws://` or `wss://` prefix.
-    pub fn server_url(&self) -> String {
-        server_url(&self.local_addr)
-    }
-}
+    fn stop_listener(&mut self) {
+        let Some(join_handle) = self.listener_join_handle.take() else {
+            return;
+        };
 
-/// Sync handle for the [`RerunServer`]
-///
-/// When dropped, the server will be shut down.
-pub struct RerunServerHandle {
-    local_addr: std::net::SocketAddr,
-    shutdown_tx: tokio::sync::broadcast::Sender<()>,
+        self.shutdown_flag
+            .store(true, std::sync::atomic::Ordering::Release);
+
+        if let Err(err) = self.poller.notify() {
+            re_log::warn!("Error notifying WebSocket server listener: {err}");
+            return;
+        }
+
+        if let Err(err) = join_handle.join() {
+            re_log::warn!("Error joining listener thread: {err:?}");
+        }
+    }
 }
 
-impl Drop for RerunServerHandle {
+impl Drop for RerunServer {
     fn drop(&mut self) {
         re_log::info!("Shutting down Rerun server on {}", self.server_url());
-        self.shutdown_tx.send(()).ok();
+        self.stop_listener();
     }
 }
 
-impl RerunServerHandle {
-    /// Create new [`RerunServer`] to relay [`LogMsg`]s to a websocket.
-    /// Returns a [`RerunServerHandle`] that will shutdown the server when dropped.
-    ///
-    /// A `bind_ip` of `"0.0.0.0"` is a good default.
-    /// A port of 0 will let the OS choose a free port.
-    ///
-    /// The caller needs to ensure that there is a `tokio` runtime running.
-    pub fn new(
-        rerun_rx: ReceiveSet<LogMsg>,
-        bind_ip: String,
-        requested_port: RerunServerPort,
-        server_memory_limit: MemoryLimit,
-    ) -> Result<Self, RerunServerError> {
-        let (shutdown_tx, shutdown_rx) = tokio::sync::broadcast::channel(1);
-
-        let rt = tokio::runtime::Handle::current();
-
-        let ws_server = rt.block_on(tokio::spawn(async move {
-            RerunServer::new(bind_ip, requested_port, server_memory_limit).await
-        }))??;
-
-        let local_addr = ws_server.local_addr;
-
-        tokio::spawn(async move {
-            ws_server
-                .listen_with_graceful_shutdown(rerun_rx, shutdown_rx)
-                .await
-        });
-
-        Ok(Self {
-            local_addr,
-            shutdown_tx,
-        })
-    }
+/// Broadcasts messages to all connected clients and stores a history of messages to resend to new clients.
+///
+/// This starts a thread which will close when the underlying `ReceiveSet` gets a quit message or looses all its connections.
+/// TODO(andreas): There should be a way to close this from the outside as well,
+/// so that on server shut down we're not left with a thread serving clients.
```

Comment:
So there's no way to cancel/shut down the server now? That sounds like a very serious limitation.

Reply:
This is quasi pre-existing, we already hat this problem before with the broadcast tokio thread.

But thinking about this again, previously the thread would dangle around but the connections were already closed (I believe), so now the connections are still open making this a regression I have to address.